crates/sui-graphql-client/src/streams.rs

```diff
@@ -0,0 +1,157 @@
+use crate::Error;
+use crate::Page;
+
+use futures::Stream;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+/// A stream that yields items from a paginated query.
+pub struct PageStream<T, F, Fut>
+where
+    F: Fn(Option<String>) -> Fut,
+    Fut: Future<Output = Result<Page<T>, Error>>,
+{
+    query_fn: F,
+    current_page: Option<Page<T>>,
+    current_future: Option<Pin<Box<Fut>>>,
+    current_index: usize,
+    finished: bool,
+}
+
+// Implement Unpin for PageStream since none of our fields need to be pinned
+impl<T, F, Fut> Unpin for PageStream<T, F, Fut>
+where
+    F: Fn(Option<String>) -> Fut,
+    Fut: Future<Output = Result<Page<T>, Error>>,
+{
+}
+
+impl<T, F, Fut> PageStream<T, F, Fut>
+where
+    F: Fn(Option<String>) -> Fut,
+    Fut: Future<Output = Result<Page<T>, Error>>,
+{
+    pub fn new(query_fn: F) -> Self {
+        Self {
+            query_fn,
+            current_page: None,
+            current_future: None,
+            current_index: 0,
+            finished: false,
+        }
+    }
+}
+
+impl<T, F, Fut> Stream for PageStream<T, F, Fut>
+where
+    T: Clone,
+    F: Fn(Option<String>) -> Fut,
+    Fut: Future<Output = Result<Page<T>, Error>>,
+{
+    type Item = Result<T, Error>;
+
+    /// Polls the stream for the next item.
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // We can now use self directly since we implemented Unpin
+        if self.finished {
+            return Poll::Ready(None);
+        }
+
+        // If we have a current page, return the next item
+        if let Some(page) = &self.current_page {
+            if self.current_index < page.data().len() {
+                let item = page.data()[self.current_index].clone();
```

Comment:
this is not ideal, but we only have a small number of elements in each vec<T>, so maybe it's fine.